File: Backend/app/services/billing_service.py
from sqlalchemy.orm import Session

from app.models.billing import Billing
from app.schemas.billing import BillingCreate, BillingUpdate


def create_bill(db: Session, bill_data: BillingCreate):
    new_bill = Billing(**bill_data.model_dump())

    db.add(new_bill)
    db.commit()
    db.refresh(new_bill)

    return new_bill


def get_all_bills(db: Session):
    return db.query(Billing).all()


def get_bill_by_id(db: Session, bill_id: int):
    return db.query(Billing).filter(
        Billing.id == bill_id
    ).first()


def update_bill(
    db: Session,
    bill_id: int,
    bill_data: BillingUpdate
):
    bill = get_bill_by_id(db, bill_id)

    if not bill:
        return None

    update_data = bill_data.model_dump(
        exclude_unset=True
    )

    for key, value in update_data.items():
        setattr(bill, key, value)

    db.commit()
    db.refresh(bill)

    return bill


def delete_bill(db: Session, bill_id: int):
    bill = get_bill_by_id(db, bill_id)

    if not bill:
        return None

    db.delete(bill)
    db.commit()

    return bill

File: frontend/src/services/billingService.js
import axios from "axios";

const API_URL = "http://127.0.0.1:8000";

export const getBills = async () => {
    const response = await axios.get(`${API_URL}/billing`);
    return response.data;
};

export const getBillById = async (billId) => {
    const response = await axios.get(`${API_URL}/billing/${billId}`);
    return response.data;
};

export const createBill = async (billData) => {
    const response = await axios.post(`${API_URL}/billing`, billData);
    return response.data;
};

export const updateBill = async (billId, billData) => {
    const response = await axios.put(
        `${API_URL}/billing/${billId}`,
        billData
    );
    return response.data;
};

export const deleteBill = async (billId) => {
    const response = await axios.delete(
        `${API_URL}/billing/${billId}`
    );
    return response.data;
};